File: src/utils/insightEngine.js
// Insight engine -- pure computed analysis, no LLM needed.
// Correlates dose log against biometrics and lab results.

const METRIC_LABELS = {
  sleep: 'sleep', steps: 'steps', heartRate: 'heart rate',
  hrv: 'HRV', bloodOxygen: 'blood oxygen', weight: 'weight',
  activeEnergy: 'active energy', exerciseMin: 'exercise',
};

function dateOf(iso) {
  return iso.slice(0, 10);
}

function mean(arr) {
  if (!arr.length) return 0;
  return arr.reduce((s, v) => s + v, 0) / arr.length;
}

function stddev(arr) {
  if (arr.length < 2) return 0;
  const m = mean(arr);
  return Math.sqrt(arr.reduce((s, v) => s + (v - m) ** 2, 0) / arr.length);
}

// For a given substance, find biometric entries on days it was taken vs. not.
// Returns { substance, metric, onMean, offMean, delta, pct, direction, confidence }
export function computeCorrelations(doseLog, biometrics) {
  if (!doseLog.length || !biometrics.length) return [];

  const insights = [];
  const substanceIds = [...new Set(doseLog.map(e => e.substanceId))];

  for (const subId of substanceIds) {
    const onDays = new Set(
      doseLog.filter(e => e.substanceId === subId).map(e => dateOf(e.timestamp))
    );

    const onVals = {};
    const offVals = {};

    for (const bio of biometrics) {
      if (!bio.metrics) continue;
      const isOn = onDays.has(bio.date);
      for (const [metric, val] of Object.entries(bio.metrics)) {
        if (val == null || val === '') continue;
        const bucket = isOn ? onVals : offVals;
        (bucket[metric] = bucket[metric] || []).push(Number(val));
      }
    }

    for (const metric of Object.keys(METRIC_LABELS)) {
      const on = onVals[metric] || [];
      const off = offVals[metric] || [];
      // Need at least 3 data points in each group for meaningful insight
      if (on.length < 3 || off.length < 3) continue;

      const onMean = mean(on);
      const offMean = mean(off);
      if (offMean === 0) continue;

      const delta = onMean - offMean;
      const pct = Math.abs(delta / offMean) * 100;
      if (pct < 5) continue; // not meaningful

      const direction = delta > 0 ? 'higher' : 'lower';
      // Confidence: based on sample sizes
      const confidence = Math.min(1, (on.length + off.length) / 20);

      insights.push({
        type: 'correlation',
        substanceId: subId,
        metric,
        metricLabel: METRIC_LABELS[metric] || metric,
        onMean: Math.round(onMean * 10) / 10,
        offMean: Math.round(offMean * 10) / 10,
        delta: Math.round(Math.abs(delta) * 10) / 10,
        pct: Math.round(pct),
        direction,
        sampleSize: on.length + off.length,
        confidence,
      });
    }
  }

  return insights.sort((a, b) => b.confidence * b.pct - a.confidence * a.pct);
}

// Detect biometric readings that deviate >2 SD from personal mean
export function anomalyAlerts(biometrics) {
  if (biometrics.length < 5) return [];

  const alerts = [];
  const metricHistory = {};

  for (const bio of biometrics) {
    if (!bio.metrics) continue;
    for (const [metric, val] of Object.entries(bio.metrics)) {
      if (val == null || val === '') continue;
      (metricHistory[metric] = metricHistory[metric] || []).push({
        date: bio.date, value: Number(val),
      });
    }
  }

  for (const [metric, history] of Object.entries(metricHistory)) {
    if (history.length < 5) continue;
    const values = history.map(h => h.value);
    const m = mean(values);
    const sd = stddev(values);
    const latest = history.sort((a, b) => b.date.localeCompare(a.date))[0];

    if (Math.abs(latest.value - m) > 2 * sd) {
      alerts.push({
        type: 'anomaly',
        metric,
        metricLabel: METRIC_LABELS[metric] || metric,
        value: latest.value,
        mean: Math.round(m * 10) / 10,
        date: latest.date,
        direction: latest.value > m ? 'above' : 'below',
        confidence: 0.85,
      });
    }
  }

  return alerts;
}

// Detect lab markers trending toward out-of-range over 3+ readings
export function detectLabTrends(labResults) {
  if (!labResults.length) return [];

  const markerHistory = {};
  for (const result of labResults) {
    for (const m of result.markers) {
      const key = m.name;
      (markerHistory[key] = markerHistory[key] || []).push({
        date: result.date, ...m,
      });
    }
  }

  const trends = [];
  for (const [name, history] of Object.entries(markerHistory)) {
    if (history.length < 3) continue;
    const sorted = history.sort((a, b) => a.date.localeCompare(b.date));
    const recent = sorted.slice(-3);

    // Check if values monotonically increasing or decreasing
    const rising = recent[0].value < recent[1].value && recent[1].value < recent[2].value;
    const falling = recent[0].value > recent[1].value && recent[1].value > recent[2].value;
    if (!rising && !falling) continue;

    const latest = recent[recent.length - 1];
    const hasRef = latest.refLow != null || latest.refHigh != null;
    if (!hasRef) continue;

    // Only flag if trending toward out-of-range
    const hi = latest.refHigh;
    const lo = latest.refLow;
    const nearHigh = hi != null && rising && latest.value > hi * 0.8;
    const nearLow = lo != null && falling && latest.value < lo * 1.2;
    if (!nearHigh && !nearLow) continue;

    trends.push({
      type: 'lab_trend',
      marker: name,
      direction: rising ? 'rising' : 'falling',
      currentValue: latest.value,
      unit: latest.unit,
      refLow: lo,
      refHigh: hi,
      readingCount: recent.length,
      confidence: 0.75,
    });
  }

  return trends;
}

// Flagged lab markers from most recent result
export function flaggedLabMarkers(labResults) {
  if (!labResults.length) return [];
  const sorted = [...labResults].sort((a, b) => b.date.localeCompare(a.date));
  const latest = sorted[0];
  return latest.markers
    .filter(m => m.flag !== 'normal')
    .map(m => ({
      type: 'lab_flag',
      marker: m.name,
      value: m.value,
      unit: m.unit,
      flag: m.flag,
      date: latest.date,
      lab: latest.lab,
      confidence: 1.0,
    }));
}

// Top 5 insights, sorted by confidence then impact
export function topInsights(doseLog, biometrics, labResults = []) {
  const all = [
    ...flaggedLabMarkers(labResults),
    ...anomalyAlerts(biometrics),
    ...detectLabTrends(labResults),
    ...computeCorrelations(doseLog, biometrics),
  ];
  return all.sort((a, b) => b.confidence - a.confidence).slice(0, 5);
}

// Human-readable summary for an insight
export function insightText(insight, getName) {
  switch (insight.type) {
    case 'lab_flag':
      return `${insight.marker} is ${insight.flag} (${insight.value} ${insight.unit}) from ${insight.lab} result on ${insight.date}`;
    case 'anomaly':
      return `${insight.metricLabel} was ${insight.direction} normal on ${insight.date} (${insight.value} vs avg ${insight.mean})`;
    case 'lab_trend':
      return `${insight.marker} has been ${insight.direction} across ${insight.readingCount} readings, approaching ${insight.direction === 'rising' ? 'upper' : 'lower'} limit`;
    case 'correlation': {
      const name = getName ? getName(insight.substanceId) : insight.substanceId;
      return `${insight.metricLabel} is ${insight.pct}% ${insight.direction} on days you take ${name}`;
    }
    default:
      return '';
  }
}

export function insightIcon(type) {
  switch (type) {
    case 'lab_flag': return 'flag';
    case 'anomaly': return 'alert';
    case 'lab_trend': return 'trend';
    case 'correlation': return 'link';
    default: return 'info';
  }
}

File: ios/Services/InsightEngine.swift
import Foundation

struct Insight: Identifiable {
    var id = UUID()
    var type: InsightType
    var title: String
    var detail: String
    var severity: InsightSeverity
}

enum InsightType: String {
    case labFlag, anomaly, labTrend, correlation
}

enum InsightSeverity {
    case info, warning, urgent

    var color: String {
        switch self {
        case .info: return "blue"
        case .warning: return "orange"
        case .urgent: return "red"
        }
    }
}

struct InsightEngine {

    // Flagged markers from most recent lab result
    static func flaggedLabMarkers(labResults: [LabResult]) -> [Insight] {
        guard let latest = labResults.sorted(by: { $0.date > $1.date }).first else { return [] }
        return latest.flaggedMarkers.map { m in
            let severity: InsightSeverity = m.flag == .critical ? .urgent : .warning
            return Insight(
                type: .labFlag,
                title: "\(m.name) is \(m.flag.label.lowercased())",
                detail: "\(m.value) \(m.unit) — from \(latest.lab) on \(formatted(latest.date))",
                severity: severity
            )
        }
    }

    // Markers trending in one direction over last 3+ readings
    static func labTrends(labResults: [LabResult]) -> [Insight] {
        guard labResults.count >= 3 else { return [] }

        var markerHistory: [String: [(date: Date, marker: LabMarker)]] = [:]
        for result in labResults {
            for m in result.markers {
                markerHistory[m.name, default: []].append((date: result.date, marker: m))
            }
        }

        var insights: [Insight] = []
        for (name, history) in markerHistory {
            let sorted = history.sorted { $0.date < $1.date }
            guard sorted.count >= 3 else { continue }
            let recent = sorted.suffix(3)
            let values = recent.map { $0.marker.value }

            let rising = values[0] < values[1] && values[1] < values[2]
            let falling = values[0] > values[1] && values[1] > values[2]
            guard rising || falling else { continue }

            let latest = recent.last!.marker
            let hi = latest.refHigh
            let lo = latest.refLow
            let nearHigh = hi != nil && rising && latest.value > hi! * 0.8
            let nearLow = lo != nil && falling && latest.value < lo! * 1.2
            guard nearHigh || nearLow else { continue }

            let direction = rising ? "rising" : "falling"
            insights.append(Insight(
                type: .labTrend,
                title: "\(name) trending \(direction)",
                detail: "\(recent.count) consecutive readings moving toward \(rising ? "upper" : "lower") limit",
                severity: .warning
            ))
        }
        return insights
    }

    // Biometric anomalies: readings > 2 SD from personal mean
    static func anomalyAlerts(biometrics: [BiometricEntry]) -> [Insight] {
        guard biometrics.count >= 5 else { return [] }

        // Extract named metric series from typed BiometricEntry fields
        typealias MetricRow = (date: Date, value: Double)
        var metricHistory: [String: [MetricRow]] = [:]

        func push(_ key: String, _ val: Double?, _ date: Date) {
            guard let v = val else { return }
            metricHistory[key, default: []].append((date: date, value: v))
        }

        for bio in biometrics {
            push("sleep", bio.sleepHours, bio.date)
            push("heartRate", bio.heartRate.map(Double.init), bio.date)
            push("hrv", bio.hrv, bio.date)
            push("bloodOxygen", bio.bloodOxygen, bio.date)
            push("weight", bio.weight, bio.date)
            push("steps", bio.steps.map(Double.init), bio.date)
        }

        let metricLabels: [String: String] = [
            "sleep": "Sleep", "steps": "Steps", "heartRate": "Heart rate",
            "hrv": "HRV", "bloodOxygen": "Blood oxygen", "weight": "Weight",
        ]

        var insights: [Insight] = []
        for (metric, history) in metricHistory {
            guard history.count >= 5 else { continue }
            let values = history.map { $0.value }
            let m = values.reduce(0, +) / Double(values.count)
            let variance = values.map { ($0 - m) * ($0 - m) }.reduce(0, +) / Double(values.count)
            let sd = sqrt(variance)
            guard sd > 0 else { continue }

            let latest = history.sorted { $0.date > $1.date }.first!
            guard abs(latest.value - m) > 2 * sd else { continue }

            let label = metricLabels[metric] ?? metric
            let direction = latest.value > m ? "above" : "below"
            insights.append(Insight(
                type: .anomaly,
                title: "\(label) unusual on \(formatted(latest.date))",
                detail: "\(Int(latest.value)) vs avg \(Int(m)) — \(direction) normal range",
                severity: .warning
            ))
        }
        return insights
    }

    // Correlation between substance and biometric
    static func correlations(doseEntries: [DoseEntry], biometrics: [BiometricEntry], getName: (DoseEntry) -> String) -> [Insight] {
        guard doseEntries.count >= 5, biometrics.count >= 5 else { return [] }

        let substanceIds = Set(doseEntries.map { $0.substanceKey })
        let metricLabels: [String: String] = [
            "sleep": "sleep", "heartRate": "heart rate", "hrv": "HRV",
        ]

        let calendar = Calendar.current
        var insights: [Insight] = []

        for subId in substanceIds {
            let onDays = Set(doseEntries.filter { $0.substanceKey == subId }
                .map { calendar.startOfDay(for: $0.timestamp) })

            var onVals: [String: [Double]] = [:]
            var offVals: [String: [Double]] = [:]

            for bio in biometrics {
                let bioDay = calendar.startOfDay(for: bio.date)
                let isOn = onDays.contains(bioDay)

                func bucket(_ key: String, _ val: Double?) {
                    guard let v = val else { return }
                    if isOn { onVals[key, default: []].append(v) }
                    else { offVals[key, default: []].append(v) }
                }
                bucket("sleep", bio.sleepHours)
                bucket("heartRate", bio.heartRate.map(Double.init))
                bucket("hrv", bio.hrv)
            }

            for (metric, label) in metricLabels {
                let on = onVals[metric] ?? []
                let off = offVals[metric] ?? []
                guard on.count >= 3, off.count >= 3 else { continue }

                let onMean = on.reduce(0, +) / Double(on.count)
                let offMean = off.reduce(0, +) / Double(off.count)
                guard offMean != 0 else { continue }

                let pct = abs((onMean - offMean) / offMean) * 100
                guard pct >= 8 else { continue }

                let direction = onMean > offMean ? "higher" : "lower"
                let entry = doseEntries.first { $0.substanceKey == subId }!
                let name = getName(entry)
                insights.append(Insight(
                    type: .correlation,
                    title: "\(label.capitalized) \(Int(pct))% \(direction) with \(name)",
                    detail: "Based on \(on.count + off.count) logged days",
                    severity: .info
                ))
            }
        }

        return insights
    }

    static func topInsights(
        doseEntries: [DoseEntry],
        biometrics: [BiometricEntry],
        labResults: [LabResult],
        getName: (DoseEntry) -> String
    ) -> [Insight] {
        let all: [Insight] = flaggedLabMarkers(labResults: labResults)
            + labTrends(labResults: labResults)
            + anomalyAlerts(biometrics: biometrics)
            + correlations(doseEntries: doseEntries, biometrics: biometrics, getName: getName)

        let order: [InsightSeverity] = [.urgent, .warning, .info]
        return all.sorted { order.firstIndex(of: $0.severity)! < order.firstIndex(of: $1.severity)! }
            .prefix(5).map { $0 }
    }

    private static func dateString(_ date: Date) -> String {
        let f = DateFormatter()
        f.dateFormat = "yyyy-MM-dd"
        return f.string(from: date)
    }

    private static func formatted(_ date: Date) -> String {
        let f = DateFormatter()
        f.dateStyle = .medium
        return f.string(from: date)
    }
}
